### DISCUSSION TITLE
Helper tool installation yields "authorizationFailure" error.

### DISCUSSION THREAD

**Discussion by mihajlovicn** [trilemma-dev/SwiftAuthorizationSample#2] [Q&A] [answered]
Dear gentlemen,

I'm attempting to run the sample application (Xcode 13.1; macOS 11.6.1) based on the current master branch.
When trying to install the Helper Tool app a "authorizationFailure" error/exception is being thrown (please see screenshot).
What could be the reason?

Personal development signing certificate has been set for both targets.

A short support message is very much appreciated.
Thank you very much in advance. ;)

<img width="666" alt="Authorization_Sample_and_SwiftAuthorizationSample_xcodeproj" src="https://user-images.githubusercontent.com/557120/141687798-57cbf191-35ea-4427-a732-fd3eee4cd6b6.png">

The log is only showing the following trace:
`2021-11-14 16:29:53.584503+0100 Swift Authorization Sample[81859:4601116] MTLIOAccelDevice bad MetalPluginClassName property (null)
2021-11-14 16:29:53.585918+0100 Swift Authorization Sample[81859:4601116] +[MTLIOAccelDevice registerDevices]: Zero Metal services found
`

Nenad

**Comment by mihajlovicn** [comment]
```
./SMJobBlessUtil.py check "/Users/nenad/Library/Developer/Xcode/DerivedData/SwiftAuthorizationSample-druzmcsaxmjyiugpscmrernvsthe/Build/Products/Debug/Swift Authorization Sample.app"
SMJobBlessUtil.py: tool designated requirement (identifier "com.cut2kill.SwiftAuthorizationApp.helper" and anchor apple generic and certificate leaf[subject.CN] = "Apple Development: xxxxxxxxxxx@xxxxxx.com (xxxxxxxxx)" and certificate 1[field.1.2.840.113635.100.6.2.1] /* exists */) doesn't match entry in 'SMPrivilegedExecutables' (anchor apple generic and identifier "com.cut2kill.SwiftAuthorizationApp.helper" and certificate leaf[subject.OU] = xxxxxxxx)
```

**Reply by jakaplan** [reply]
This by itself shouldn't be a problem. While they intentionally don't match, that doesn't mean the code signing requirement in SMPrivilegedExecutables doesn't meet the designated requirement for the helper tool. For details see: https://github.com/trilemma-dev/SwiftAuthorizationSample/blob/f0c291165ff79d0ab6aa50acd16e095a4f39e8e6/BuildScripts/PropertyListModifier.swift#L87-L129

**Comment by jakaplan** [comment]
What happens if you make an Archive build and then "distribute" it as a Developer ID? Does that resulting build also have this issue?

There's probably something going wrong with the code signing, but just from the information you provided I can't tell what it is. If it works when you create an Archive build that might help narrow it down.

**Reply by mihajlovicn** [reply]
Thank you, Josh.

I'm going to try that with the distribution build, this is a good plan. I currently do not have a Developer plan enrolled for signing/provisioning "Release" builds but this should normally not be required for regular Test runs via Xcode - but maybe this is a problem for  SMPrivilegedExecutables.

![SwiftAuthorizationSample_xcodeproj](https://user-images.githubusercontent.com/557120/141783158-dc6dbd29-6f1a-4dd1-b484-787a5acf8943.png)

Greetings to NZ, I wish I could live there. :)

**Reply by jakaplan** [reply]
Ah that may indeed be the problem. If it turns out this works after you do so, let me know and I'll update the README documentation to reflect this. Apple definitely did not fully document the requirements for `SMJobBless` (for example there are completely undocumented update eligibility rules based on the helper tool's bundle version number).

If this _doesn't_ work, let me know and I can look into add a menu item to the sample which will display a bunch of diagnostic information. It'd likely take me several days to have the time to work on that.

I've never been to Vienna, I've heard it's lovely!

**Comment by mihajlovicn** [comment]
I have some findings now.

The `PropertyListModifier` build script creates the following `SMPrivilegedExecutables` and  `SMAuthorizedClients` entries in my case:

```
=== SwiftAuthorziationApp ===

<key>SMPrivilegedExecutables</key>
<dict>
    <key>com.example.SwiftAuthorizationApp.helper</key>
    <string>anchor apple generic and identifier "com.example.SwiftAuthorizationApp.helper" and certificate leaf[subject.OU] = 8xxxxxxx48</string>
</dict>
```

```
==== Helper Tool ====

<key>SMAuthorizedClients</key>
<array>
    <string>anchor apple generic and identifier "com.example.SwiftAuthorizationApp" and info[CFBundleVersion] &gt;= "1.0.0" and certificate leaf[subject.OU] = 8xxxxxxx48</string>
</array>
```

This does not work in my case for whatever reason (see initial problem description).
I'm not enrolled in a Developer plan for actual provisioning of release stuff. I have a Team ID for for automatic signing/provisioning of "Development" builds.

After running the `SMJobBlessUtil.py setreq` command from the original apple example, the plist entries get "fixed" to:

```
=== SwiftAuthorziationApp ===

<key>SMPrivilegedExecutables</key>
<dict>
    <key>com.example.SwiftAuthorizationApp.helper</key>
    <string>identifier "com.example.SwiftAuthorizationApp.helper" and anchor apple generic and certificate leaf[subject.CN] = "Apple Development: xxxxxxx@xxxx.at (xxxxxxxx)" and certificate 1[field.1.2.840.113635.100.6.2.1] /* exists */</string>
</dict>
```
and:

```
==== Helper Tool ====

<key>SMAuthorizedClients</key>
<array>
    <string>identifier "com.example.SwiftAuthorizationApp" and anchor apple generic and certificate leaf[subject.CN] = "Apple Development: xxxxxxx@xxxx.at (xxxxxxxx)" and certificate 1[field.1.2.840.113635.100.6.2.1] /* exists */</string>
</array>
```

This change in the plists allows me to install the privileged helper tool via the example app successfully.
What is the obvious difference?:
The `SMJobBlessUtil.py setreq` uses the "Common Name" from my Apple Development certificate while the OOTB `func organizationalUnitRequirement()` extracts the "Organisational Unit". I know that the comments of the method extensively elaborate why the "OU" should be sufficient but at least in my case it doesn't work.

@jakaplan, don't feel obliged to "fix" this or so if you do not see some need for your personal self gain (= understanding). It would be cool to understand if this is a general problem or only specific to me. The Archive build did not make a difference.

Best regards,
Nenad

**Comment by jakaplan** [comment]
@mihajlovicn If you update to the latest commit and run the sample there will now be a "Print Diagnostic Info" option in the menu bar. Mind running that and pasting what it prints out? It will contain the information for the app and helper tool used for the code signing requirements.

**Reply by mihajlovicn** [reply]
Hi Josh,

this is the output:

```
==============This App=============
CFBundleIdentifier: com.example.SwiftAuthorizationApp
CFBundleVersion: 1.0.0
Common Name: Apple Development: nenad.mihajlovic@chello.at (5MK4Y8QK93)
Organizational Unit: 8AU2ACQZ48

========Bundled Helper Tool========
CFBundleIdentifier: com.example.SwiftAuthorizationApp.helper
CFBundleVersion: 1.0.1
Common Name: Apple Development: nenad.mihajlovic@chello.at (5MK4Y8QK93)
Organizational Unit: 8AU2ACQZ48

=======Installed Helper Tool=======
Helper tool is not installed
```

And the relevant plist content after the `SMJobBlessUtil.py setreq` "fix":

```
<key>SMPrivilegedExecutables</key>
	<dict>
		<key>com.example.SwiftAuthorizationApp.helper</key>
		<string>identifier "com.example.SwiftAuthorizationApp.helper" and anchor apple generic and certificate leaf[subject.CN] = "Apple Development: nenad.mihajlovic@chello.at (5MK4Y8QK93)" and certificate 1[field.1.2.840.113635.100.6.2.1] /* exists */</string>
	</dict>

<key>SMAuthorizedClients</key>
	<array>
		<string>identifier "com.example.SwiftAuthorizationApp" and anchor apple generic and certificate leaf[subject.CN] = "Apple Development: nenad.mihajlovic@chello.at (5MK4Y8QK93)" and certificate 1[field.1.2.840.113635.100.6.2.1] /* exists */</string>
	</array>
```

OOTB:

```
<key>SMAuthorizedClients</key>
	<array>
		<string>anchor apple generic and identifier "com.example.SwiftAuthorizationApp" and info[CFBundleVersion] &gt;= "1.0.0" and certificate leaf[subject.OU] = 8AU2ACQZ48</string>
	</array>

<key>SMPrivilegedExecutables</key>
	<dict>
		<key>com.example.SwiftAuthorizationApp.helper</key>
		<string>anchor apple generic and identifier "com.example.SwiftAuthorizationApp.helper" and certificate leaf[subject.OU] = 8AU2ACQZ48</string>
	</dict>
```

**Reply by jakaplan** [reply]
The information that printed out looks correct. The only thing that looks slightly suspicious is `info[CFBundleVersion] &gt;= "1.0.0"`. I'm wondering if somehow the `>` is actually being encoded wrong or it's just whatever you're using to print it out is decoding it wrong.

In order to disambiguate if this is a code signing issue vs some issue with your macOS installation, could you try the following:
 - Using the sample app you already have, uninstall the helper tool
 - In a terminal run `sudo security -q authorizationdb remove com.example.SwiftAuthorizationSample.secure-action`
 - Download and run [this sample](https://github.com/trilemma-dev/SwiftAuthorizationSample/files/7579037/Swift.Authorization.Sample.app.zip) which I've code signed (and notarized), let me know if you're able to successfully install the helper tool

I'd very much like to get to the bottom of this and resolve this issue for you (and likely others in the future).

**Reply by mihajlovicn** [reply]
I've performed the steps above. I was able to successfully install the helper tool via your example app. The `>` is just a encoding issue.

**Reply by mihajlovicn** [reply]
If I remove the entry `and certificate leaf[subject.OU] = 8AU2ACQZ48` from both plists, everything works fine. (This is defying the purpose). I'm going to re-install Xcode and delete all my Developer certificates now.

**Reply by mihajlovicn** [reply]
@jakaplan, no change after re-installing Xcode and the developer certificates.

Executed `codesign -d -vvv com.example.SwiftAuthorizationApp.helper` on your helper and then on mine, after removing  `and certificate leaf[subject.OU] = 8AU2ACQZ48`:

@jakaplan:
```
Executable=/Library/PrivilegedHelperTools/com.example.SwiftAuthorizationApp.helper
Identifier=com.example.SwiftAuthorizationApp.helper
Format=Mach-O universal (x86_64 arm64)
CodeDirectory v=20500 size=4052 flags=0x10000(runtime) hashes=115+7 location=embedded
Hash type=sha256 size=32
CandidateCDHash sha256=070da4e2b6ba08765d52893d9adce528a3677871
CandidateCDHashFull sha256=070da4e2b6ba08765d52893d9adce528a3677871ddc4667b5cd84bd079bd3127
Hash choices=sha256
CMSDigest=070da4e2b6ba08765d52893d9adce528a3677871ddc4667b5cd84bd079bd3127
CMSDigestType=2
CDHash=070da4e2b6ba08765d52893d9adce528a3677871
Signature size=8975
Authority=Developer ID Application: Joshua Kaplan (R96J7HJPH8)
Authority=Developer ID Certification Authority
Authority=Apple Root CA
Timestamp=22.11.2021 at 08:33:24
Info.plist entries=5
TeamIdentifier=R96J7HJPH8
Runtime Version=12.0.0
Sealed Resources=none
Internal requirements count=1 size=232
```

@mihajlovicn:
```
Executable=/Library/PrivilegedHelperTools/com.example.SwiftAuthorizationApp.helper
Identifier=com.example.SwiftAuthorizationApp.helper
Format=Mach-O thin (x86_64)
CodeDirectory v=20400 size=12588 flags=0x0(none) hashes=382+7 location=embedded
Hash type=sha256 size=32
CandidateCDHash sha256=a52d035b0e9c5c9f209047a430bc79525c4a125a
CandidateCDHashFull sha256=a52d035b0e9c5c9f209047a430bc79525c4a125a27c0ca80fb6bef2c4cf287f8
Hash choices=sha256
CMSDigest=a52d035b0e9c5c9f209047a430bc79525c4a125a27c0ca80fb6bef2c4cf287f8
CMSDigestType=2
CDHash=a52d035b0e9c5c9f209047a430bc79525c4a125a
Signature size=4798
Authority=Apple Development: nenad.mihajlovic@chello.at (5MK4Y8QK93)
Authority=Apple Worldwide Developer Relations Certification Authority
Authority=Apple Root CA
Signed Time=23.11.2021 at 15:00:32
Info.plist entries=5
TeamIdentifier=8AU2ACQZ48
Sealed Resources=none
Internal requirements count=1 size=216
```

There are some interesting differences but I have no clue if they are the reason for the problem or not.

**Reply by mihajlovicn** [reply]
![Keychain_Access](https://user-images.githubusercontent.com/557120/143045962-fc2aebfe-8aa8-4ec9-a12b-7709bc251f90.png)

Are you having a similar structure of certificates in your keychain?

**Reply by mihajlovicn** [reply]
Hm... I've created a completely new Apple ID and used it for Signing.
Result: It works now.

I also played around with the Xcode 13.2 2 beta but only the new Apple ID seemed to help.

**Reply by jakaplan** [reply]
Glad it works for you now, although all the more perplexing. Mind running `codesign` once again so we can compare the output from what you posted previously?

**Reply by mihajlovicn** [reply]
Here we go, `codesign` results with the new Apple ID:

```
Executable=/Library/PrivilegedHelperTools/com.example.SwiftAuthorizationApp.helper
Identifier=com.example.SwiftAuthorizationApp.helper
Format=Mach-O thin (x86_64)
CodeDirectory v=20500 size=12596 flags=0x10000(runtime) hashes=382+7 location=embedded
Hash type=sha256 size=32
CandidateCDHash sha256=01b8d8f2deeac2461e23b3e8e294c8bf5f9d894e
CandidateCDHashFull sha256=01b8d8f2deeac2461e23b3e8e294c8bf5f9d894ea6cd9ba737382217a81aea61
Hash choices=sha256
CMSDigest=01b8d8f2deeac2461e23b3e8e294c8bf5f9d894ea6cd9ba737382217a81aea61
CMSDigestType=2
CDHash=01b8d8f2deeac2461e23b3e8e294c8bf5f9d894e
Signature size=4792
Authority=Apple Development: nmradosin@gmail.com (LKRX5V7NVR)
Authority=Apple Worldwide Developer Relations Certification Authority
Authority=Apple Root CA
Signed Time=26.11.2021 at 08:12:05
Info.plist entries=5
TeamIdentifier=N9VJP85R3H
Runtime Version=12.0.0
Sealed Resources=none
Internal requirements count=1 size=208
```

```
Executable=/Users/nenad/Library/Developer/Xcode/DerivedData/SwiftAuthorizationSample-druzmcsaxmjyiugpscmrernvsthe/Build/Products/Debug/Swift Authorization Sample.app/Contents/MacOS/Swift Authorization Sample
Identifier=com.example.SwiftAuthorizationApp
Format=bundle with Mach-O thin (x86_64)
CodeDirectory v=20500 size=13293 flags=0x10000(runtime) hashes=404+7 location=embedded
Hash type=sha256 size=32
CandidateCDHash sha256=98b95d9692c56d2c7f270a00c4d342e5e078387b
CandidateCDHashFull sha256=98b95d9692c56d2c7f270a00c4d342e5e078387b487ec168a220ef8a2b60650c
Hash choices=sha256
CMSDigest=98b95d9692c56d2c7f270a00c4d342e5e078387b487ec168a220ef8a2b60650c
CMSDigestType=2
CDHash=98b95d9692c56d2c7f270a00c4d342e5e078387b
Signature size=4792
Authority=Apple Development: nmradosin@gmail.com (LKRX5V7NVR)
Authority=Apple Worldwide Developer Relations Certification Authority
Authority=Apple Root CA
Signed Time=26.11.2021 at 08:12:06
Info.plist entries=20
TeamIdentifier=N9VJP85R3H
Runtime Version=12.0.0
Sealed Resources version=2 rules=13 files=5
Internal requirements count=1 size=204
```

**Reply by mihajlovicn** [reply]
And, once more with the not working Apple ID (i renamed the old one):

```
Executable=/Users/nenad/Library/Developer/Xcode/DerivedData/SwiftAuthorizationSample-druzmcsaxmjyiugpscmrernvsthe/Build/Products/Debug/com.example.SwiftAuthorizationApp.helper
Identifier=com.example.SwiftAuthorizationApp.helper
Format=Mach-O thin (x86_64)
CodeDirectory v=20500 size=12596 flags=0x10000(runtime) hashes=382+7 location=embedded
Hash type=sha256 size=32
CandidateCDHash sha256=0032f530106c239617b64c5a9d9cd57b7cd5b2c7
CandidateCDHashFull sha256=0032f530106c239617b64c5a9d9cd57b7cd5b2c7eb4d494263f2e864a32eb517
Hash choices=sha256
CMSDigest=0032f530106c239617b64c5a9d9cd57b7cd5b2c7eb4d494263f2e864a32eb517
CMSDigestType=2
CDHash=0032f530106c239617b64c5a9d9cd57b7cd5b2c7
Signature size=4798
Authority=Apple Development: nenadmihajlovic@chello.at (2S2UUR6SUU)
Authority=Apple Worldwide Developer Relations Certification Authority
Authority=Apple Root CA
Signed Time=26.11.2021 at 08:18:46
Info.plist entries=5
TeamIdentifier=6XJ28CNRF7
Runtime Version=12.0.0
Sealed Resources=none
Internal requirements count=1 size=216
```

```
Executable=/Users/nenad/Library/Developer/Xcode/DerivedData/SwiftAuthorizationSample-druzmcsaxmjyiugpscmrernvsthe/Build/Products/Debug/Swift Authorization Sample.app/Contents/MacOS/Swift Authorization Sample
Identifier=com.example.SwiftAuthorizationApp
Format=bundle with Mach-O thin (x86_64)
CodeDirectory v=20500 size=13293 flags=0x10000(runtime) hashes=404+7 location=embedded
Hash type=sha256 size=32
CandidateCDHash sha256=96551a6eb1defb26b99ac28463bb8eae1bc01f91
CandidateCDHashFull sha256=96551a6eb1defb26b99ac28463bb8eae1bc01f91e738fd5c87c435621b67ae30
Hash choices=sha256
CMSDigest=96551a6eb1defb26b99ac28463bb8eae1bc01f91e738fd5c87c435621b67ae30
CMSDigestType=2
CDHash=96551a6eb1defb26b99ac28463bb8eae1bc01f91
Signature size=4798
Authority=Apple Development: nenadmihajlovic@chello.at (2S2UUR6SUU)
Authority=Apple Worldwide Developer Relations Certification Authority
Authority=Apple Root CA
Signed Time=26.11.2021 at 08:18:49
Info.plist entries=20
TeamIdentifier=6XJ28CNRF7
Runtime Version=12.0.0
Sealed Resources version=2 rules=13 files=5
Internal requirements count=1 size=212
```

Let me know if you're able to spot something here.

**Reply by jakaplan** [reply]
Thanks for doing that. Unfortunately nothing looks inconsistent between them in a meaningful way.

If you're up for trying out one last thing, it'd be good to know if `codesign` itself thinks the requirement is met for the app and helper tool signed with your older certificate. To do so run: `codesign -vvvv -R="certificate leaf[subject.OU]=6XJ28CNRF7" <path to app/helper tool>`

If it satisfies the requirement the last line it prints out will look something like: `<path to app/helper tool> explicit requirement satisfied`. Otherwise it'll look like: `test-requirement: code failed to satisfy specified code requirement(s)`

If you're comfortable doing so, could you upload your self-signed app bundle (just zip it up first) with the original certificate? I'll take a deeper look as my time permits. I have a few other theories, but they're all increasingly less likely and don't want to use up too much of your time particularly since it's now working for you.

**Reply by amomchilov** [reply]
This could be worth contacting apple about, or at least posting on their forums.

**Reply by mihajlovicn** [reply]
Hm... interesting:

With the old certificate where it's not working:
```
codesign -vvvv -R="certificate leaf[subject.OU]=6XJ28CNRF7" Swift\ Authorization\ Sample.app
error: invalid or corrupted code requirement(s)
Requirement syntax error(s):
line 1:29: unexpected token: =
line 1:30: expecting EOF, found '6'

codesign -vvvv -R="certificate leaf[subject.OU]=6XJ28CNRF7" com.example.SwiftAuthorizationApp.helper
error: invalid or corrupted code requirement(s)
Requirement syntax error(s):
line 1:29: unexpected token: =
line 1:30: expecting EOF, found '6'
```

With the new certificate/new Developer ID:
```
codesign -vvvv -R="certificate leaf[subject.OU]=N9VJP85R3H" Swift\ Authorization\ Sample.app
Swift Authorization Sample.app: valid on disk
Swift Authorization Sample.app: satisfies its Designated Requirement
Swift Authorization Sample.app: explicit requirement satisfied

codesign -vvvv -R="certificate leaf[subject.OU]=N9VJP85R3H" com.example.SwiftAuthorizationApp.helper
com.example.SwiftAuthorizationApp.helper: valid on disk
com.example.SwiftAuthorizationApp.helper: satisfies its Designated Requirement
com.example.SwiftAuthorizationApp.helper: explicit requirement satisfied
```

@jakaplan , send me an ping mail to "nmradosin@gmail.com" and i'll respond with the zip.
By self-signed app bundle you mean the ziped content at: `/Users/xxxxx/Library/Developer/Xcode/DerivedData/SwiftAuthorizationSample-druzmcsaxmjyiugpscmrernvsthe/Build/Products/Debug`?

**Reply by jakaplan** [reply]
Have sent you an email. I don't need the everything that in `Debug` folder, just the `Swift Authorization Sample.app` is sufficient. All I meant was because an `app` bundle is a folder not a file, it'll need to be zipped in order to send it.

**Reply by mihajlovicn** [reply]
> Have sent you an email. I don't need the everything that in `Debug` folder, just the `Swift Authorization Sample.app` is sufficient. All I meant was because an `app` bundle is a folder not a file, it'll need to be zipped in order to send it.

Sure Josh, got the point. I've created a archive build, zipped it and sent it out.

**Comment by amomchilov** [comment]
Just to chime in: I don't have a paid Apple developer account either, but it still works just fine for me.

**Reply by mihajlovicn** [reply]
hi Alexander,

thank you for supporting. ;)

can you maybe show your setup of the Signing configuration and Build Settings in this regard?

![SwiftAuthorizationSample_xcodeproj](https://user-images.githubusercontent.com/557120/142861261-85d9954e-b1db-4f79-9644-7d492a5efc9d.png)

![SwiftAuthorizationSample_xcodeproj](https://user-images.githubusercontent.com/557120/142861493-879257e4-45eb-4908-8013-643e2c96c87c.png)

BR, Nenad

**Reply by amomchilov** [reply]
I used the "levels" view, but here it is:

![Screen Shot 2021-11-22 at 12 35 57 PM](https://user-images.githubusercontent.com/5703449/142908820-9635bb5b-e07c-410d-9e9f-fb679286a465.png)
![Screen Shot 2021-11-22 at 12 37 13 PM](https://user-images.githubusercontent.com/5703449/142908822-a559b167-7c95-41ba-b514-e84b8d45c4f6.png)

**Comment by jakaplan** [comment, marked as answer]
I'm pretty sure I figured out what was wrong and fixed it. See #8 for details.

**Reply by mihajlovicn** [reply]
Super, i'll give it a try this week.
Thank you very much.

**Reply by mihajlovicn** [reply]
yes, it works now OOTB. ;)
thank you very much.
